### PR TITLE
RFC - not for merging - alternative uniform handling

### PR DESCRIPTION
**Pull request by aardgoose**
I've been looking at #8552, #9927 and the issues around fixing instances / interactive / gpu single material example.

This is a prototype of the program pulling uniforms from the renderer rather than hard coding the requirements in webGLRenderer setProgram().

I've kept it opt in  ( Material.isExperimentalMaterial = true ) to allow easy testing without breaking existing code. The new uniform values are called parameters in the code to prevent confusion with existing uniform handling code.

Features:

- ShaderMaterial uniforms can be handled as normal material properties such as color, opacity etc. myMaterial.parameter = x; rather than myMaterial.uniforms.parameter.value = x;

- Parameters can be pulled from a hierarchy of sources ( implemented object -> material in prototype).
- Parameters are versioned, no needsUpdate required at object/material/parameter.
( changes to object values eg THREE.Color() are also handled by updating a parameter version on get as well as set operations). versioning is transparent to the user.
- Only changed uniforms are sent to the GPU.
- If extended to built in materials could remove a lot of the custom per material uniform setting in setProgram.

Obviously this is very raw and things like cloning/copying parameters are not implemented.

Is this worth exploring further, say converting some built in materials to this method?
Any interest in this proposal, comments? 

This PR also includes fixing the instances / interactive /gpu example which removes the need for the removed dynamic Uniform or onBeforeRender() callbacks (which don't work for this case anyway) by using object specific parameters.

### CODE REVIEW

**Comment by fallenoak** [comment]
I'll try give this a spin later this weekend.

From your description, it sounds like a pretty comprehensive set of improvements to the issues with the existing uniform system.

**Comment by aardgoose** [comment]
OK, a I counldn't resist.

LineBasic, MeshBasic, MeshLambert, MeshStandard & MeshPhysical materials converted over and working for most examples.  MeshPhong and PointsMaterial require mechanism to adjust parameters before setting uniforms. Fog needs putting in here to.

Existing code path can be easily restored by removing the .isExperimentalMaterial flag.

File: src/materials/MeshLambertMaterial.js [hunk 1]
```diff
@@ -46,28 +46,30 @@ function MeshLambertMaterial( parameters ) {
 
 	this.type = 'MeshLambertMaterial';
 
-	this.color = new Color( 0xffffff ); // diffuse
+	this.isExperimentalMaterial = true;
 
-	this.map = null;
+	this.addParameter( 'color', new Color( 0xffffff ), 'diffuse' ); // diffuse
```
Comment (mrdoob) [inline src/materials/MeshLambertMaterial.js]: Maybe a less intrusive approach could be something like... `this.defineParameters( this, [ 'color', 'lightMap', 'lightMapIntensity', ... ] );

**Comment by aardgoose** [comment]
Could do, however some names are aliased eg color -> diffuse. As an alternative it could parse through UniformsLib.<shader>.uniforms and add an .alias.  and handle ShaderMaterials in a similar way, almost making built in and custom shaders modular (apart from things like DEFINES which are another issue). That could remove a lot of code from built in materials and stop the duplication in Material objects and UniformsLibs.  

Is this whole approach interesting? It certainly makes uniforms a lot more dynamic (overrides by Object etc) and creates a much simplified setProgram, and avoids all the "material.uniform.<name>.value".

**Comment by mrdoob** [comment]
> Is this whole approach interesting? It certainly makes uniforms a lot more dynamic (overrides by Object etc) and creates a much simplified setProgram, and avoids all the "material.uniform..value".

Yeah, the current approach clearly needs some re-architecting. Keep in mind serialisation and the structure you're creating when doing so though.

**Comment by aardgoose** [comment]
Ok, this is more sound I think,  dropped versioning - too much js performance overhead, and also simplified things a lot.  Nearly examples work with no change with built in materials, RectArea lighting doesn't work because the pre-computed textures are not built in yet. Very little material specific knowledge in the renderer required now.

**Comment by arose** [comment]
Is this ready for some testing? I would really like to see something like this included so that dynamic updating of uniforms can be done again.

**Comment by aardgoose** [comment]
minor merge fix to keep in line with dev. 
AFAICS all examples work correctly.

As code it works. The interactive / instances / gpu examples demonstrates per object uniforms but the onBeforeRender handler could equally be used. Or even addDerivedParameter() gives you something equivalent to the old dynamic uniform.
 
I haven't looked at light uniforms, I think any work there should take into account the possibility of using UBOs in WebGL2 for lighting uniforms.  I'd certainly be interested in any results.

**Comment by aardgoose** [comment]
Ok, latest merge conflicts fixed.

@mrdoob any thoughts?

**Comment by aardgoose** [comment]
Dropping this one, much simplified version to come,

**Pull request closed without merging**